### ISSUE TITLE
cli: validate --environments should not test repetitively the same image

### ISSUE THREAD

**Issue by tiborsimko** [reanahub/reana-client#487]
WRT environment validation, one observation.  For BSM search, the output is huge, see below. This is because the same image is validated many times depending on where in the workflow it appears. Is there a change to quickly add a  sort-and-unique operation on extracted image names and tags, to have the simplest list to check, in order not to check the same image many times?

If it is fast, let's add it before 0.7.3 is out.  If it needs refactoring, let's create new issue for 0.7.4 (that could also have the #484 included).

```console
$ reana-demo-bsm-search
$ reana-client validate --environments
==> Verifying REANA specification file... /home/tibor/private/project/reana/src/reana-demo-bsm-search/reana.yaml
  -> SUCCESS: Valid REANA specification file.
==> Verifying workflow parameters and commands... 
  -> WARNING: Yadage input parameter "weightvariations" does not seem to be used.
  -> WARNING: Yadage input parameter "mcname" does not seem to be used.
  -> WARNING: Yadage input parameter "mcweight" does not seem to be used.
  -> WARNING: Yadage input parameter "shapevars" does not seem to be used.
==> Verifying environments in REANA specification file...
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists locally.
  -> SUCCESS: Environment image reanahub/reana-env-root6:6.18.04 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 has the correct format.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists locally.
  -> SUCCESS: Environment image reanahub/reana-demo-bsm-search:1.0.0 exists in Docker Hub.
  -> WARNING: Environment image UID is recommended to be 1000. UID 0 was found.
```

_Originally posted by @tiborsimko in https://github.com/reanahub/reana-client/pull/480#r600695305_

**Comment by mvidalgarcia** [comment]
An idea to optimize the output repetition/granularity:

> ATM we're spiting warnings as we find them, while traversing workflow steps and so on. In order to do these optimizations in the output I think we should accumulate all the output messages (warnings, success and errors) in a dictionary of sets (to avoid rep), and once the validation is done, print all the information as we wish. This also will help to improve things like the granularity of messages mentioned [here](https://github.com/reanahub/blog.reana.io/pull/11#discussion_r600321100)

**Comment by mvidalgarcia** [comment]
Closed by https://github.com/reanahub/reana-client/pull/507
